server/src/test/java/org/elasticsearch/index/snapshots/blobstore/SlicedInputStreamTests.java

```diff
@@ -138,6 +140,38 @@ protected InputStream openSlice(int slice) throws IOException {
         streamsOpened.forEach(stream -> assertTrue(stream.closed));
     }
 
+    public void testRandomMarkResetInBigSlice() throws IOException {
+        AtomicReference<IncreasingBytesUnlimitedInputStream> stream = new AtomicReference<>();
+        SlicedInputStream input = new SlicedInputStream(1) {
+            @Override
+            protected InputStream openSlice(int slice) throws IOException {
+                assertThat(slice, equalTo(0));
+                stream.set(new IncreasingBytesUnlimitedInputStream());
+                return stream.get();
+            }
+        };
+
+        // Skip up to a random point that is larger than 4GiB so that the marked offset is larger than an int (ES-9639).
+        final long mark = ByteSizeValue.ofGb(randomIntBetween(5, 6)).getBytes() + randomLongBetween(1, 1024);
+        input.skip(mark);
```

Comment:
Could we also implement `skip` on `SlicedInputStream` rather than reading several GiB of simulated data byte-by-byte? This test takes multiple seconds to run right now.